### ISSUE TITLE
[BUG] Optimizer weights output appears to be hugely dependent on learning rate

### ISSUE THREAD

**Issue by 1over137** [open-spaced-repetition/fsrs4anki#113]
**Describe the bug**
I fed my data to the optimizer, and to my surprise, it gave me this:
`w = [1.953, 2.0356, 4.9726, -0.8455, -0.793, 0.0203, 1.596, -0.0252, 1.0496, 1.8532, -0.3544, 0.5807, 1.2157];`
which produced rather huge intervals (10+ days from first graduating step!)
I went back and adjusted some parameters, and it seems like reducing the optimizer learning rate greatly reduces the first two values.
At learning rate 2e-4, I get
`[1.4371, 1.4612, 4.9139, -0.6709, -0.671, 0.0254, 1.5635, -0.0152, 0.994, 1.925, -0.2799, 0.5231, 1.0741]`
And at learning rate 5e-5, I get
`[1.1196, 1.1241, 4.9365, -0.5343, -0.5525, 0.1436, 1.4813, -0.01, 0.8912, 1.9765, -0.2248, 0.5022, 1.0156]`
Is this the expected behavior?

Increasing epoch count also doesn't make them converge.

**Comment by L-M-Sherlock** [comment]
How many review logs do you have? The size of the dataset and the iteration of training also have a huge impact on the output.

**Comment by L-M-Sherlock** [comment]
By the way, could you paste the analysis generated from the optimizer of your review log here?

**Comment by 1over137** [comment]
I have quite a lot more reviews, but I decided to cut off the ones before mid 2021 since the review methods I used before then were rather uneven and contained a lot of heterogeneous materials. `revlog.csv` has about 44000 reviews. 

Analysis? do you mean this:
```
  r_history  avg_interval  avg_retention  stability  factor  group_cnt
          1           1.7          0.765        1.0     inf       7978
        1,3           3.9          0.876        4.3    4.30       4155
      1,3,3           8.6          0.883        9.2    2.14       2684
    1,3,3,3          17.8          0.857       13.8    1.50       1483
  1,3,3,3,3          37.0          0.812       19.4    1.41        606
1,3,3,3,3,3          77.1          0.708       23.1    1.19        128
          2           1.0          0.901        1.1     inf        234
        2,3           3.2          0.943        6.3    5.73        154
          3           1.5          0.962        5.4     inf       9070
        3,3           3.9          0.966       15.2    2.81       6527
      3,3,3           9.0          0.960       23.5    1.55       5036
    3,3,3,3          18.6          0.941       35.2    1.50       3052
  3,3,3,3,3          39.5          0.914       46.9    1.33       1423
3,3,3,3,3,3          74.3          0.853       55.6    1.19        411
          4           3.8          0.966       12.1     inf      11436
        4,3           8.1          0.975       38.9    3.21       7367
      4,3,3          18.0          0.963       57.7    1.48       5147
    4,3,3,3          34.0          0.947       77.2    1.34       2525
  4,3,3,3,3          46.3          0.906       50.1    0.65        452
```

**Comment by 1over137** [comment]
As a side note, I wonder, is it possible to filter out data from cards that have been removed or suspended from the deck? I mainly studied cards from shared decks or card generation tools, so a lot of them are too easy or badly made, so I delete/suspend a lot of them, sometimes only after a review or two.

**Comment by L-M-Sherlock** [comment]
> 4           3.8          0.966       12.1     inf      11436

The data shows that it needs 12 days to achieve 90% retention. So 10+ days from the first graduating step seems normal if you press easy in the first review.

**Comment by L-M-Sherlock** [comment]
> As a side note, I wonder, is it possible to filter out data from cards that have been removed or suspended from the deck? I mainly studied cards from shared decks or card generation tools, so a lot of them are too easy or badly made, so I delete/suspend a lot of them, sometimes only after a review or two.

You can export the filtered deck with some conditions.

**Comment by 1over137** [comment]
When I used deck export instead of collection export, I get this:
```
  r_history  avg_interval  avg_retention  stability  factor  group_cnt
          1           1.0          0.855        0.7     inf        622
        1,3           3.4          0.876        2.8    4.00        497
      1,3,3           8.4          0.871        7.7    2.75        370
    1,3,3,3          17.8          0.883       19.7    2.56        204
          3           1.0          0.947        6.2     inf       3476
        3,3           3.6          0.955       10.9    1.76       3058
      3,3,3          10.5          0.953       23.5    2.16       2716
    3,3,3,3          29.8          0.959       78.8    3.35       2319
  3,3,3,3,3          82.3          0.927      116.3    1.48       1721
3,3,3,3,3,3         178.4          0.821       88.5    0.76        233
```
Which gave a different set of weights at lr=5e-4
`[1.7837, 1.8422, 4.9525, -0.8357, -0.7816, 0.0143, 1.6159, -0.0133, 1.0774, 1.8984, -0.3062, 0.5664, 1.2457]`

I'll use it for now, and report back about whether the reviews are too easy or too hard in a few week's time.

**Comment by L-M-Sherlock** [comment]
> Which gave a different set of weights at lr=5e-4

The smaller the lr, the smaller the change in the weight.

**Comment by L-M-Sherlock** [comment]
> I'll use it for now, and report back about whether the reviews are too easy or too hard in a few week's time.

I am looking forward to your report. It is useful for me to improve the optimizer.

**Comment by L-M-Sherlock** [comment]
> I'll use it for now, and report back about whether the reviews are too easy or too hard in a few week's time.

Do you have any reports? Any feedback is welcome.

**Comment by 1over137** [comment]
I would be happy to make a report, but currently I lack the time to actually analyze the data in detail.
Would you be willing to make a simple jupyter notebook to evaluate retention statistics? Maybe add it to the repo so more people can evaluate their performance

My impression overall is that these weights worked reasonably well for my Russian vocabulary deck. Anki statistics says 82% retention for young cards and 85%, which is a bit lower than the requestRetention 0.86, but totally acceptable given that I'm doing about 50% less reviews than I would be doing on Anki algorithm. On the other hand, the retention rate on my Japanese deck fell off a cliff, and I had to set weights to the default in order to stop forgetting 50% of the cards. For context, my level of Russian is much higher than Japanese, which could contribute to the cards being easier. 

I would suggest recommending training separate weights for different decks, in case people are studying material with varying difficulty.

**Comment by L-M-Sherlock** [comment]
> Would you be willing to make a simple jupyter notebook to evaluate retention statistics? Maybe add it to the repo so more people can evaluate their performance

The optimizer has been able to generate the analysis of retention in Section 1.2 Create time-series feature & analysis.

![image](https://user-images.githubusercontent.com/32575846/216752739-793b12ea-2776-435f-a632-90ed987ced48.png)

**Comment by L-M-Sherlock** [comment]
> Overall, thank you for saving what basically amounts to a few hours of my life over the past month alone.

I'm glad to hear that! I hope there will be more people using FSRS to save time.

**Comment by L-M-Sherlock** [comment]
> On the other hand, the retention rate on my Japanese deck fell off a cliff, and I had to set weights to the default in order to stop forgetting 50% of the cards.

You can upload your Japanese deck to the optimizer for generating the adaptive parameters.

**Comment by 1over137** [comment]
> > Would you be willing to make a simple jupyter notebook to evaluate retention statistics? Maybe add it to the repo so more people can evaluate their performance
> 
> The optimizer has been able to generate the analysis of retention in Section 1.2 Create time-series feature & analysis.
> 
> ![image](https://user-images.githubusercontent.com/32575846/216752739-793b12ea-2776-435f-a632-90ed987ced48.png)

I tried that, but it doesn't work. It only gives me two rows.
```
r_history  avg_interval  avg_retention  stability  factor  group_cnt
        1           7.9          0.700        2.4     inf        345
        3           9.0          0.898        8.9     inf        353
```

**Comment by 1over137** [comment]
Oh, seems like I need to reduce the minimum counts
here it is:
```
r_history  avg_interval  avg_retention  stability  factor  group_cnt
        1           7.9          0.700        2.4     inf        345
      1,3          14.0          0.833        8.1    3.38         30
        2           9.0          0.843        5.6     inf         51
        3           9.0          0.898        8.9     inf        353
```
I have no idea why the first row says average interval is 7.9 days. I believe the first interval if i pressed 1 should be 2 days.

**Comment by 1over137** [comment]
Also, I don't understand what "i" is in revlog_history.csv and I don't see it explained anywhere. Can you explain what it is?

**Comment by L-M-Sherlock** [comment]
> I tried that, but it doesn't work. It only gives me two rows.

It means that the data of reviews is not enough to generate more stats.

**Comment by L-M-Sherlock** [comment]
> I have no idea why the first row says average interval is 7.9 days. I believe the first interval if i pressed 1 should be 2 days.

It considers the delays. Maybe some reviews after you pressed "again" were delayed?

**Comment by L-M-Sherlock** [comment]
> Also, I don't understand what "i" is in revlog_history.csv and I don't see it explained anywhere. Can you explain what it is?

The i means the i-th review.

**Comment by 1over137** [comment]
> > I have no idea why the first row says average interval is 7.9 days. I believe the first interval if i pressed 1 should be 2 days.
> 
> It considers the delays. Maybe some reviews after you pressed "again" were delayed?

Unlikely. I review my cards every day with only 1-2 lapses in the past month. Shouldn't stretch the average to 8 days. Maybe it could be the suspended cards but they are not being recalled, and I shouldn't have ever pressed again on most of them.

**Comment by L-M-Sherlock** [comment]
> I have no idea why the first row says average interval is 7.9 days. I believe the first interval if i pressed 1 should be 2 days.

Could you share the uploaded file with me? I will check it.

**Comment by 1over137** [comment]
Here's the revlog
[revlog.csv](https://github.com/open-spaced-repetition/fsrs4anki/files/10693422/revlog.csv)

**Comment by L-M-Sherlock** [comment]
<html  xmlns:o="urn:schemas-microsoft-com:office:office"  xmlns:x="urn:schemas-microsoft-com:office:excel"  xmlns="http://www.w3.org/TR/REC-html40">
<head>

<style>

</style>

</head>
<body>



id | cid | usn | r | ivl | last_lvl | factor | time | type | create_date | review_date | real_days | delta_t | i | r_history | t_history
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
1674597177419 | 1620264313393 | 3140 | 1 | -600 | 9 | 1300 | 6918 | 1 | 2021-05-06 09:25:13+08:00 | 2023-01-25 05:52:57+08:00 | 2459969.5 | 9 | 2 | 1 | 0
1673825598962 | 1620264313777 | 3127 | 3 | 35 | 14 | 1300 | 4344 | 1 | 2021-05-06 09:25:13+08:00 | 2023-01-16 07:33:18+08:00 | 2459960.5 | 14 | 2 | 1 | 0
1673749954272 | 1620264314325 | 3125 | 3 | 24 | 13 | 1300 | 3991 | 1 | 2021-05-06 09:25:14+08:00 | 2023-01-15 10:32:34+08:00 | 2459959.5 | 13 | 2 | 1 | 0
1673040920450 | 1620264314418 | 3110 | 3 | 9 | 5 | 1300 | 2889 | 1 | 2021-05-06 09:25:14+08:00 | 2023-01-07 05:35:20+08:00 | 2459951.5 | 5 | 2 | 1 | 0
1673910353825 | 1620264314638 | 3127 | 3 | 56 | 16 | 1300 | 2193 | 1 | 2021-05-06 09:25:14+08:00 | 2023-01-17 07:05:53+08:00 | 2459961.5 | 16 | 2 | 1 | 0
1675291539957 | 1620264314671 | 3151 | 3 | 26 | 12 | 1300 | 2629 | 1 | 2021-05-06 09:25:14+08:00 | 2023-02-02 06:45:39+08:00 | 2459977.5 | 12 | 2 | 1 | 0
1674252083487 | 1620264314736 | 3137 | 3 | 24 | 10 | 1300 | 3414 | 1 | 2021-05-06 09:25:14+08:00 | 2023-01-21 06:01:23+08:00 | 2459965.5 | 10 | 2 | 1 | 0
1674511707483 | 1620264314908 | 3140 | 3 | 41 | 22 | 1300 | 1845 | 1 | 2021-05-06 09:25:14+08:00 | 2023-01-24 06:08:27+08:00 | 2459968.5 | 22 | 2 | 1 | 0
1674081546999 | 1620264315042 | 0 | 3 | 32 | 17 | 1300 | 1422 | 1 | 2021-05-06 09:25:15+08:00 | 2023-01-19 06:39:06+08:00 | 2459963.5 | 17 | 2 | 1 | 0
1673568173340 | 1620264315089 | 3122 | 3 | 21 | 11 | 1300 | 5664 | 1 | 2021-05-06 09:25:15+08:00 | 2023-01-13 08:02:53+08:00 | 2459957.5 | 11 | 2 | 1 | 0
1672956021078 | 1620264315140 | 3109 | 3 | 7 | 4 | 1300 | 3429 | 1 | 2021-05-06 09:25:15+08:00 | 2023-01-06 06:00:21+08:00 | 2459950.5 | 4 | 2 | 1 | 0
1673825583178 | 1620264315173 | 3127 | 3 | 49 | 14 | 1300 | 6788 | 1 | 2021-05-06 09:25:15+08:00 | 2023-01-16 07:33:03+08:00 | 2459960.5 | 14 | 2 | 1 | 0
1673750159022 | 1620264315305 | 3125 | 3 | 47 | 13 | 1300 | 2894 | 1 | 2021-05-06 09:25:15+08:00 | 2023-01-15 10:35:59+08:00 | 2459959.5 | 13 | 2 | 1 | 0
1673302375565 | 1620264315329 | 3117 | 1 | -600 | 8 | 1300 | 1558 | 1 | 2021-05-06 09:25:15+08:00 | 2023-01-10 06:12:55+08:00 | 2459954.5 | 8 | 2 | 1 | 0
1674511803385 | 1620264315540 | 3140 | 3 | 58 | 23 | 1300 | 2150 | 1 | 2021-05-06 09:25:15+08:00 | 2023-01-24 06:10:03+08:00 | 2459968.5 | 23 | 2 | 1 | 0
1675291332208 | 1620264315841 | 3151 | 3 | 39 | 11 | 1300 | 6666 | 1 | 2021-05-06 09:25:15+08:00 | 2023-02-02 06:42:12+08:00 | 2459977.5 | 11 | 2 | 1 | 0
1673393297111 | 1620264315844 | 3119 | 3 | 20 | 9 | 1300 | 1935 | 1 | 2021-05-06 09:25:15+08:00 | 2023-01-11 07:28:17+08:00 | 2459955.5 | 9 | 2 | 1 | 0
1673302419800 | 1620264316028 | 3117 | 2 | 10 | 8 | 1300 | 7690 | 1 | 2021-05-06 09:25:16+08:00 | 2023-01-10 06:13:39+08:00 | 2459954.5 | 8 | 2 | 1 | 0
1673302431202 | 1620264316258 | 3117 | 2 | 10 | 8 | 1300 | 11374 | 1 | 2021-05-06 09:25:16+08:00 | 2023-01-10 06:13:51+08:00 | 2459954.5 | 8 | 2 | 1 | 0
1673223817653 | 1620264316329 | 3114 | 3 | 14 | 7 | 1300 | 2153 | 1 | 2021-05-06 09:25:16+08:00 | 2023-01-09 08:23:37+08:00 | 2459953.5 | 7 | 2 | 1 | 0
1673910556603 | 1620264316562 | 3127 | 3 | 20 | 8 | 1300 | 1783 | 1 | 2021-05-06 09:25:16+08:00 | 2023-01-17 07:09:16+08:00 | 2459961.5 | 8 | 2 | 1 | 0
1673998063174 | 1620264316662 | 3128 | 3 | 17 | 8 | 1300 | 2450 | 1 | 2021-05-06 09:25:16+08:00 | 2023-01-18 07:27:43+08:00 | 2459962.5 | 8 | 2 | 1 | 0
1673393425914 | 1620264316725 | 3119 | 3 | 33 | 9 | 1310 | 3002 | 1 | 2021-05-06 09:25:16+08:00 | 2023-01-11 07:30:25+08:00 | 2459955.5 | 9 | 2 | 1 | 0
1673479076609 | 1620264316750 | 3120 | 3 | 33 | 9 | 1310 | 2764 | 1 | 2021-05-06 09:25:16+08:00 | 2023-01-12 07:17:56+08:00 | 2459956.5 | 9 | 2 | 1 | 0
1673479116070 | 1620264316752 | 3120 | 1 | -600 | 9 | 1300 | 1101 | 1 | 2021-05-06 09:25:16+08:00 | 2023-01-12 07:18:36+08:00 | 2459956.5 | 9 | 2 | 1 | 0
1673393315675 | 1620264316754 | 3119 | 1 | -600 | 9 | 1300 | 5665 | 1 | 2021-05-06 09:25:16+08:00 | 2023-01-11 07:28:35+08:00 | 2459955.5 | 9 | 2 | 1 | 0
1673393309985 | 1620264316755 | 3119 | 1 | -600 | 9 | 1300 | 11293 | 1 | 2021-05-06 09:25:16+08:00 | 2023-01-11 07:28:29+08:00 | 2459955.5 | 9 | 2 | 1 | 0
1673393383844 | 1620264316756 | 3119 | 3 | 33 | 9 | 1310 | 15255 | 1 | 2021-05-06 09:25:16+08:00 | 2023-01-11 07:29:43+08:00 | 2459955.5 | 9 | 2 | 1 | 0



</body>

</html>

**Comment by L-M-Sherlock** [comment]
These cards' first rating is 1 (again), but their intervals are larger than 2 days.

**Comment by L-M-Sherlock** [comment]
Could you search for the card in your Anki's browser? cid:1620264313393

**Comment by 1over137** [comment]
I checked some of these. Not so sure what's happening here. Is it possible that I pressed "Easy" at some point on the same day after pressing "Again"? I never missed reviews, and max review limit is infinity, so cards should never really be overdue.

**Comment by L-M-Sherlock** [comment]
> I checked some of these. Not so sure what's happening here. Is it possible that I pressed "Easy" at some point on the same day after pressing "Again"? I never missed reviews, and max review limit is infinity, so cards should never really be overdue.

<html  xmlns:o="urn:schemas-microsoft-com:office:office"  xmlns:x="urn:schemas-microsoft-com:office:excel"  xmlns="http://www.w3.org/TR/REC-html40">
<head>
</head>
<body>


<html  xmlns:o="urn:schemas-microsoft-com:office:office"  xmlns:x="urn:schemas-microsoft-com:office:excel"  xmlns="http://www.w3.org/TR/REC-html40">
<head>


id | cid | usn | r | ivl | last_lvl | factor | time | type
-- | -- | -- | -- | -- | -- | -- | -- | --
1673825678316 | 1620264313393 | 3127 | 1 | -600 | 32 | 1300 | 8116 | 1
1673826010258 | 1620264313393 | 3127 | 1 | -600 | -600 | 1300 | 5890 | 2
1673826136235 | 1620264313393 | 3127 | 3 | 9 | -600 | 1300 | 904 | 2
1674597177419 | 1620264313393 | 3140 | 1 | -600 | 9 | 1300 | 6918 | 1



</body>

</html>

For example, this card's interval given by Anki is 9 days.

**Comment by L-M-Sherlock** [comment]
Oh, maybe I know what happened. Do you change the value of `revlog_start_date`?

**Comment by 1over137** [comment]
No idea what happened here then. Does that mean FSRS is giving it 9 days? Using card info seems to not give any information. 
My current weights for the deck are: `[1.7684, 1.8516, 4.9533, -0.9402, -0.7572, 0.0162, 1.6288, -0.0126, 1.0872, 1.9024, -0.3084, 0.5743, 1.2249]`
If memory serves me right, the card is probably first seen a long time ago but recently forgotten and relearned. Maybe it did not get included in revlog_start_date.

**Comment by L-M-Sherlock** [comment]
> No idea what happened here then. Does that mean FSRS is giving it 9 days? Using card info seems to not give any information.

FSRS with your current weights doesn't gives 9 days for those cards when their first rating is 'again'. You can check it in the optimizer's 'Preview' section.

**Comment by 1over137** [comment]
Now that I'm using a deck export, it shouldn't be including any deleted notes, so I'll run the optimizer with revlog_start_date to much earlier again. However, this includes a bunch of earlier reviews, where I pressed Good for new cards a lot because the algorithm did not care about the first rating and gives the same interval regardless. That's why I'm setting revlog_start_date to 2023-01-01.

**Comment by L-M-Sherlock** [comment]
> Now that I'm using a deck export, it shouldn't be including any deleted notes, so I'll run the optimizer with revlog_start_date to much earlier again. However, this includes a bunch of earlier reviews, where I pressed Good for new cards a lot because the algorithm did not care about the first rating and gives the same interval regardless. That's why I'm setting revlog_start_date to 2023-01-01.

Thanks for this report! I find out the problem. I will fix it in the next version.

**Comment by 1over137** [comment]
```
1:again, 2:hard, 3:good, 4:easy

    r_history  avg_interval  avg_retention  stability  factor  group_cnt
            1           2.0          0.824        2.0     inf        886
          1,3           3.7          0.878        3.2    1.60        499
        1,3,3          10.4          0.884        9.2    2.87        392
      1,3,3,3          30.1          0.879       23.7    2.58        211
    1,3,3,3,3          78.0          0.768       30.1    1.27         58
  1,3,3,3,3,3         122.7          0.667       31.9    1.06          9
            2           7.0          0.873        5.5     inf         80
          2,3           1.6          0.900        2.0    0.36         18
            3           1.0          0.946        2.5     inf       5090
          3,3           3.7          0.955        8.5    3.40       4248
        3,3,3          10.2          0.959       25.8    3.04       3914
      3,3,3,3          27.4          0.965       81.6    3.16       3206
    3,3,3,3,3          70.6          0.948      134.8    1.65       2469
  3,3,3,3,3,3         146.0          0.909      147.7    1.10        565
3,3,3,3,3,3,3         286.7          0.763      114.6    0.78         54
            4           4.7          0.886        4.2     inf         17
Analysis saved!
```
For reference this is what I get when I let is see the entire deck's history. You can see 3 has the same stability as 1, and less than 2, because I pressed Good for new cards.

**Comment by Expertium** [comment]
I decided to try different learning rates and here's what I got for one of my decks
var w = [0.7415, 0.7579, 5.3359, -0.8683, -0.7972, 0.0512, 1.2051, -0.1084, 0.6051, 1.7442, -0.4295, 0.7154, 1.0661];	lr=5e-4
var w = [0.2048, 0.7932, 3.9768, -2.4059, -2.4454, 0.1143, 1.6082, -0.1525, 1.0097, 1.8227, -0.557, 0.7766, 1.8704];	lr=5e-3
I don't know which one works better in the sense of minimizing the loss function.

**Comment by L-M-Sherlock** [comment]
The optimizer has added an evaluation section in the end of notebook.

**Comment by L-M-Sherlock** [comment]
![image](https://user-images.githubusercontent.com/32575846/218240112-84f5a6e3-a6ef-44c8-839d-33eb743144bb.png)

@1over137  You can modify the learning rate and then check the evaluation. The less the loss, the more accurate the model.

This issue is off-topic a lot, so I close this issue. You can open a new issue if you are still stuck in it.

**Comment by 1over137** [comment]
I actually wrote some code to compare several of the weights trained at different learning rates and found one with the lowest loss, which is the one I mentioned above. As the results show, it seems to work quite well.
